Complete the BIP155 implementation and upgrade to TORv3

vasild opened the pull request:
This PR contains the two remaining commits from #19031 to complete the [BIP155](https://github.com/bitcoin/bips/blob/master/bip-0155.mediawiki) implementation:

`net: CAddress & CAddrMan: (un)serialize as ADDRv2`
`net: advertise support for ADDRv2 via new message`

plus one more commit:

`tor: make a TORv3 hidden service instead of TORv2`

vasild committed:
net: CAddress & CAddrMan: (un)serialize as ADDRv2

Change the serialization of `CAddrMan` to serialize its addresses
in ADDRv2/BIP155 format by default. Introduce a new `CAddrMan` format
version (3).

Add support for ADDRv2 format in `CAddress` (un)serialization.

Co-authored-by: Carl Dong <contact@carldong.me>

vasild committed:
net: advertise support for ADDRv2 via new message

Introduce a new message `sendaddrv2` to signal support for ADDRv2.
Send the new message immediately after sending the `VERACK` message.

Add support for receiving and parsing ADDRv2 messages.

Send ADDRv2 messages (instead of ADDR) to a peer if he has
advertised support for it.

Co-authored-by: Carl Dong <contact@carldong.me>

vasild committed:
tor: make a TORv3 hidden service instead of TORv2

TORv2 is deprecated [1], thus whenever we create the hidden service
ourselves create a TORv3 one instead.

[1] https://blog.torproject.org/v2-deprecation-timeline

vasild commented:
This builds on top of #19031, adding one more commit to complete the TORv3 puzzle. The dependency to #19031 is because if we create a TORv3 hidden service before #19031 is merged we will not be able to advertise it by gossiping. So, this PR is draft until all of #19031 is merged. Available for early review and testing.

jonatack commented on src/torcontrol.cpp:
```diff
@@ -464,10 +466,23 @@ TorController::TorController(struct event_base* _base, const std::string& _targe
+    if (private_key.substr(0, strlen(torv2_key_prefix)) == torv2_key_prefix) {
+        auto obsolete_key_file = GetPrivateKeyFile().string() + "_obsolete_torv2";
+        LogPrint(BCLog::TOR, "tor: Detected an obsolete TORv2 private key, will move it "
+                             "to %s and will create new TORv3 one\n", obsolete_key_file);
```
Linter fails on missing newline here.
```
$ test/lint/lint-logs.sh
All calls to LogPrintf() and LogPrint() should be terminated with \n
```

vasild replied:
Deleted this code altogether.

jonatack reviewed (commented):
Concept ACK, am running this right now.

Debug build is clean, all tests are passing for me locally, and above all, for the first time I am operating a torv3 bitcoind service.

jonatack commented on src/protocol.h:
```diff
@@ -76,6 +76,19 @@ extern const char* VERACK;
 extern const char* ADDR;
+/**
+ * The addrv2 message relays connection information for peers on the network just
+ * like the addr message, but is extended to allow gossipping of longer node
```
nit: gossiping

vasild replied:
Fixed.

jonatack commented on src/protocol.h:
```diff
@@ -76,6 +76,19 @@ extern const char* VERACK;
+ * The sendaddrv2 message signals support for receiving ADDRv2 messages (BIP155).
+ * It also implies that its sender can encode as ADDRv2 and would send ADDRv2
+ * instead of ADDR to a peer who has indicated ADDRv2 support (has sent
+ * SENDADDRv2 himself).
```
s/who has indicated ADDRv2 support (has sent SENDADDRV2 himself)/that has signaled ADDRv2 support by sending SENDADDRv2./

vasild replied:
Changed.

jonatack reviewed (commented):
Initial (light) pass through the code looks good apart from the minor comments that follow. For the three comments with the word "Sneak", suggest "Bitwise OR the" instead.

prayank23 commented:
Can we also add the option to delete the `onion_private_key` and `onion_private_key_obsolete_torv2`automatically every time bitcoind is restarted? So that it gets new address each time.

Or is it something out of scope for this PR and will need a new issue or PR to discuss?

sipa commented:
@prayank23 It can take a (very) long time for a node's rumoured address to be picked up, especially Tor ones. Getting a new one every time you restart your node is probably a good way to never get any incoming connections at all.

Saibato commented:
@prayank23
>  Can we also add the option to delete `onion_private_key` and `onion_private_key_obsolete_torv2`automatically every time bitcoind is restarted? So that it gets new address each time.

> 
tyi; #19635 we probably will need that in some form for v3 anyway.

jonatack commented:
Thinking about migration, as a follow-up before release we may need a temporary configuration option to use tor v2.

jonatack commented:
Thinking about migration, as a follow-up before release we may need a temporary configuration option to use tor v2/v3. 

Perhaps begin with v3 as opt-in, then in the following release look at making it the default.

laanwj commented on doc/files.md:
```diff
@@ -57,6 +57,7 @@ Subdirectory       | File(s)               | Description
 `./`               | `guisettings.ini.bak` | Backup of former [GUI settings](#gui-settings) after `-resetguisettings` option is used
 `./`               | `mempool.dat`         | Dump of the mempool's transactions
 `./`               | `onion_private_key`   | Cached Tor onion service private key for `-listenonion` option
+`./`               | `onion_private_key_obsolete_torv2` | Obsolete Tor onion service private key used to create TORv2 hidden services. This file is automatically created as a backup when new TORv3 private key is generated and saved into `onion_private_key`. Bitcoin Core does not need it and will never attempt to read it. Remove if you do not need the old TORv2 hidden service.
```
I'd slightly prefer creating a new file for v3 instead of moving the old one. This makes it easier to go back and forth to test this.

vasild replied:
Indeed. Done!

laanwj commented:
@prayank23 

> Can we also add the option to delete onion_private_key and onion_private_key_obsolete_torv2automatically every time bitcoind is restarted? So that it gets new address each time.

Yes, out of scope here. There's an issue for this already: #17491.

Saibato commented:
@laanwj 
> I'd slightly prefer creating a new file for v3 instead 

Feel free to grab there, then i can close that PR 

https://github.com/bitcoin/bitcoin/pull/19485

tryphe commented:
utACK 9c9ff0a8c40421b43922a6b211f99b6aaf94a798. Light code review, code looks great, apart from several uses of `char name[1025]` which seem a bit clunky. Will need to fire up a test build. Nice work!

Sjors commented:
Concept ACK, I shared Wladimir's preference for "creating a new file for v3 instead of moving the old one". Plus it'll make this change even smaller.

> Perhaps begin with v3 as opt-in, then in the following release look at making it the default.

Given that v2 is deprecated I don't think v3 needs to be opt-in, at least not for new users.

Existing users might rely on a static v2 address, e.g. because they have another node connected to it, or they paired it with some remote control software. Afaik this only applies to bitcoind users, not QT (unless @luke-jr implemented pairing in Knots?). So this change needs some emphasis in the release notes.

We could add an (instantly deprecated) option `-listenonion_v2`.

vasild commented:
Rebased on latest #19031 and addressed suggestions.

Indeed creating a new file for the new TORv3 private key seems like a better idea. It also opens up a possibility for the users to continue using the old TORv2 service if needed by copying the contents of the old file to the new file.

I think Bitcoin Core automatically creating just one Tor service is enough (that will be v3). If more complicated setup is needed, then the users can create static services in the Tor daemon - e.g. it is possible to create 5 TORv2 services and 7 TORv3 services all pointing to the same Bitcoin Core.

jonatack commented:
Practical testing of various Tor configuration options by and between @vasild and I today. Will review this weekend.

Sjors reviewed (commented):
tACK 1550aa82659506ef89db67e482d01f014f6a1977 just the last commit

I was able launch a hidden v3 service on Ubuntu and connect to it.

laanwj commented:
> I think Bitcoin Core automatically creating just one Tor service is enough (that will be v3). 

The only worry here would be that old nodes will run out of Tor nodes to connect to. But I don't know. I'm okay with this, it seems there's enough hurry to transition here that backwards compatibility concerns like that are mostly moot.

laanwj commented on doc/files.md:
```diff
@@ -56,7 +56,8 @@ Subdirectory       | File(s)               | Description
 `./`               | `guisettings.ini.bak` | Backup of former [GUI settings](#gui-settings) after `-resetguisettings` option is used
 `./`               | `mempool.dat`         | Dump of the mempool's transactions
-`./`               | `onion_private_key`   | Cached Tor onion service private key for `-listenonion` option
+`./`               | `onion_private_key`   | Obsolete: cached Tor onion service private key for `-listenonion` option. Was used for TORv2 services. Bitcoin Core does not need this file and ignores it starting from version 0.21.
```
Maybe move this line to `## Legacy subdirectories and files`, if we're no longer going to spin up a v2 service.

vasild replied:
Done.

laanwj commented on src/crypto/common.h:
```diff
@@ -53,6 +53,13 @@ void static inline WriteLE64(unsigned char* ptr, uint64_t x)
+{
+    uint16_t x;
+    memcpy((char*)&x, ptr, 2);
+    return be16toh(x);
```
Good, looks ilke we already have an implementation of `be16toh` in `src/compat/endian.h`, so it doesn't need to be added.

jonatack commented on src/protocol.h:
```diff
@@ -76,6 +76,18 @@ extern const char* VERACK;
 extern const char* ADDR;
+/**
+ * The addrv2 message relays connection information for peers on the network just
+ * like the addr message, but is extended to allow gossiping of longer node
```
99047b07 nit, maybe "gossiping of longer node addresses up to 512 bytes"

vasild replied:
I deliberately leave numbers outside of comments - as too much unnecessary details and also could go out of sync if the code is updated and the comment not.

jonatack commented on doc/files.md:
```diff
@@ -56,7 +56,8 @@ Subdirectory       | File(s)               | Description
 `./`               | `guisettings.ini.bak` | Backup of former [GUI settings](#gui-settings) after `-resetguisettings` option is used
 `./`               | `mempool.dat`         | Dump of the mempool's transactions
-`./`               | `onion_private_key`   | Cached Tor onion service private key for `-listenonion` option
+`./`               | `onion_private_key`   | Obsolete: cached Tor onion service private key for `-listenonion` option. Was used for TORv2 services. Bitcoin Core does not need this file and ignores it starting from version 0.21.
```
1550aa8 nits, s/TORv2/Tor v2/ and s/does not need/no longer needs/ (though if you move it to the legacy section, use the "replaced by" format instead)

vasild replied:
Done - "replaced by ..."

jonatack reviewed (commented):
Tested ACK 1550aa82659506ef89db67e482d01f014f6a1977 (what an appropriate commit hash for BIP155). Reviewed the changes, and over the past 10 days have done multiple debug builds, test runs, and have been running and re-running bitcoind on and off with this branch as a Tor v3 service connected to several peers that are also test-deploying this PR live on mainnet.

Thanks for adding the unit tests. A few non-blocking comments follow.

jonatack commented on src/net_processing.cpp:
```diff
@@ -2637,6 +2665,11 @@ void PeerManager::ProcessMessage(CNode& pfrom, const std::string& msg_type, CDat
     }
 
+    if (msg_type == NetMsgType::SENDADDRv2) {
+        pfrom.m_wants_addrv2 = true;
```
Had a look at if this needs a lock; it doesn't seem so.

jonatack commented:
> The only worry here would be that old nodes will run out of Tor nodes to connect to. But I don't know. I'm okay with this, it seems there's enough hurry to transition here that backwards compatibility concerns like that are mostly moot.

Maybe it would be good for the release notes to describe the situation with the planned full removal of v2 by the Tor Project and warn users running older nodes with an onion service that they will have to upgrade within less than a year after the release to v0.21+. It might not need to be in this PR though.

vasild commented:
> > I think Bitcoin Core automatically creating just one Tor service is enough (that will be v3).
> 
> ... old nodes will run out of Tor nodes to connect to ...

If there are `N` old nodes each one of them will be able to connect to the other `N-1`. Only the very last one (when `N=1`) will have nobody to connect to. Also, statically configured Tor v2 services (via `torrc`) will keep operating as long as the Tor network supports them.

According to https://bitnodes.io/nodes/ 37% are running the latest version (0.20.0 or 0.20.1), now, about 2 months before the release of 0.21.

vasild commented:
Rebase on latest #19031

vasild commented:
Address review suggestions.

vasild commented on src/torcontrol.cpp:
```diff
@@ -533,7 +533,7 @@ void TorController::auth_cb(TorControlConnection& _conn, const TorControlReply&
 
         // Finally - now create the service
         if (private_key.empty()) // No private key, generate one
```
Done

vasild commented:
Address review suggestion.

vasild commented on doc/release-notes.md:
```diff
@@ -74,6 +74,14 @@ P2P and network changes
+  removed if not needed. Bitcoin Core will not attempt to read it anymore. The
+  private key for the TORv3 service will be saved in `onion_private_key_torv3`.
+  If using the old TORv2 service is desirable (not recommended), then
+  `onion_private_key` can be copied over `onion_private_key_torv3`.
```
Done.

`cp -f onion_private_key onion_private_key_torv3` is clearer? :)

jonatack replied:
LGTM

vasild commented on src/addrman.h:
```diff
@@ -318,8 +329,15 @@ friend class CAddrManTest;
-        s << nVersion;
+        const auto stream_version_orig = s.GetVersion();
+
+        if (m_format >= Format::V3) {
```
AddrMan already had versions on its on-disk database with numbers. So I left the convention as is and just added new version "3" since the last one used was "2". This has nothing to do with "addrv2" or "torv3" :)

`IsAddrV2Format()` would spoil the strict ordering of the AddrMan on-disk versions.

jonatack replied:
Yes, I understand. Helper functions that replace the various `m_format >= Format::V<n>` may possibly be nice, but opinions can vary and it isn't a blocker.

vasild commented on src/net_processing.cpp:
```diff
@@ -2584,9 +2587,29 @@ void PeerManager::ProcessMessage(CNode& pfrom, const std::string& msg_type, CDat
+            Misbehaving(pfrom.GetId(), 20,
+                strprintf(
+                    "Supplied us with a message that contains an unparsable address: %s",
+                    e.what()));
```
Yes! (since the last push with updated tests)

> has there been discussion on how much to add to the misbehavior score for this?

I don't think so. It is currently punishing the peer as much as if he sent us an addr message with >1000 addresses. I do not insist on that.

jonatack replied:
Nice work on the tests!

vasild commented on src/net_processing.cpp:
```diff
@@ -2612,6 +2635,11 @@ void PeerManager::ProcessMessage(CNode& pfrom, const std::string& msg_type, CDat
 
+            // Skip invalid addresses, maybe a new BIP155 network id from the future.
+            if (!addr.IsValid()) {
+                continue;
```
Yes! (since the last push with updated tests)

jonatack replied:
Should skipping invalid BIP155 addresses be applied only for `addrv2` messages?
```cpp
if (!addr.IsValid() && msg_type == NetMsgType::ADDRv2) {
    continue;
}
```

jonatack commented:
Ready to move this PR out of Draft status? 😀

vasild commented:
Opened for review. This PR supersedes #19031 and completes the TORv3 support.

hebasto commented:
Concept ACK.

hebasto commented on src/addrman.h:
```diff
@@ -318,8 +329,15 @@ friend class CAddrManTest;
+            s.SetVersion(stream_version_orig | ADDRV2_FORMAT);
+        }
+
+        s << (uint8_t)m_format;
```
e025121c1ca4b86296c6b36386786522bd79a6e1

```suggestion
        s << static_cast<uint8_t>(m_format);
```

vasild replied:
Done

jonatack replied:
Per https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es49-if-you-must-use-a-cast-use-a-named-cast and *C++ Programming Language 4th Edition*, section 11.5, it might be good to update the next line as well if you retouch.
```diff
         s << static_cast<uint8_t>(m_format);
-        s << ((unsigned char)32);
+        s << static_cast<unsigned char>(32);
```

vasild replied:
That line is unmodified by the patch. I try to avoid sneaking unrelated improvements to surrounding code in order to ease reviewers and future history observers.

jonatack replied:
That's fair.

hebasto commented on src/addrman.h:
```diff
@@ -376,8 +396,17 @@ friend class CAddrManTest;
+
+        uint8_t format_tmp;
+        s >> format_tmp;
+        m_format = (Format)format_tmp;
```
e025121c1ca4b86296c6b36386786522bd79a6e1
```suggestion
        m_format = static_cast<Format>(format_tmp);
```

vasild replied:
Done

hebasto reviewed (changes requested):
Approach ACK ff76c3444c574b14a2c62999b82cba812816884b.

nit: It seems more readable if all comparisons with `m_format` will use the same sign, say `>=`, instead of mixing `>=` and `>`.

vasild commented:
Addressed suggestions by @hebasto - always compare `m_format` with `>=` and use `static_cast` instead of C-style cast.

vasild commented:
#### Addresses with length > 512 or addresses from one of the BIP155 founding networks (1..6) with wrong length (e.g. IPv4 address with length 5):

We throw an exception during unserialize which means that subsequent addresses in the same addrv2 message will not be read and the peer is `Misbehaving()`ed.

#### Addresses from unknown network id (0 or >= 7), with address length <= 512:

Those are possibly legit addresses from a future extension of BIP155. We consume the address from the stream and keep parsing subsequent addresses in the same addrv2 message. The unknown address will however not be relayed or added to addrman (it is unserialized as all-0s IPv6 address which is `!IsValid()`).

#### I2P and CJDNS addresses (BIP155 network ids 5 and 6):

[Relevant IRC discussion on how to handle those](https://bitcoincore.reviews/19845#l-319). There are concerns whether we should relay or add them to addrman because we can't further verify their validity or check if a node actually exists at that address.

The current code actually would [not add such addresses to addrman](https://github.com/bitcoin/bitcoin/blob/1769828684f16b53e5fbf65173f508b9ea1b4b9c/src/net_processing.cpp#L2626-L2628) because they are `!IsReachable()`, so we would not get "garbage" in addrman.

The current code would however relay such addresses (to 1.5 peers like other `!IsReachable()` addresses).
* This has the downside of garbage gossip traffic. Is it any worse than somebody gossiping garbage IPv6 addresses? Honest nodes don't have a way to verify if there is a node listening on a I2P address (yet). But even now we don't verify if there is a node listening on an address before relaying it.
* The benefit is that once I2P support is added in version 0.2X then older nodes, e.g. 0.21 will relay I2P addresses and help with propagation.

<details>
<summary>The change to not relay I2P and CJDNS addresses is easy, click here.</summary>

```diff
--- i/src/net_processing.cpp
+++ w/src/net_processing.cpp
@@ -2643,14 +2643,14 @@ void PeerManager::ProcessMessage(CNode& pfrom, const std::string& msg_type, CDat
             pfrom.AddAddressKnown(addr);
             if (m_banman && (m_banman->IsDiscouraged(addr) || m_banman->IsBanned(addr))) {
                 // Do not process banned/discouraged addresses beyond remembering we received them
                 continue;
             }
             bool fReachable = IsReachable(addr);
-            if (addr.nTime > nSince && !pfrom.fGetAddr && vAddr.size() <= 10 && addr.IsRoutable())
-            {
+            if (addr.nTime > nSince && !pfrom.fGetAddr && vAddr.size() <= 10 &&
+                addr.IsRoutable() && !addr.IsI2P() && !addr.IsCJDNS()) {
                 // Relay to a limited number of other nodes
                 RelayAddress(addr, fReachable, m_connman);
             }
             // Do not store addresses outside our network
             if (fReachable)
                 vAddrOk.push_back(addr);
```
</details>

What are your thoughts?

laanwj commented:
@vasild 

> Addresses with length > 512 or addresses from one of the BIP155 founding networks (1..6) with wrong length (e.g. IPv4 address with length 5):
> We throw an exception during unserialize which means that subsequent addresses in the same addrv2 message will not be read and the peer is Misbehaving()ed.

Sounds like correct behavior to me. Larger addresses are explicitly forbidden, and can be seen as an indication that the entire message is malformed.

> Addresses from unknown network id (0 or >= 7), with address length <= 512:
> We consume the address from the stream and keep parsing subsequent addresses in the same addrv2 message. The unknown address will however not be relayed or added to addrman

Seems good  to me too.

> The current code would however relay such addresses (to 1.5 peers like other !IsReachable() addresses).

I'm not sure about this. I'm not sure we should do anything (store, relay) with addresses that we cannot connect to ourselves.
But as you say this is a broader scope decision. Staying (at least in this PR) with the current behaviour for valid but unroutable addresses doesn't sound unreasonable to me.

jonatack commented on doc/files.md:
```diff
@@ -98,6 +98,7 @@ Path           | Description | Repository notes
 `blkindex.dat` | Blockchain index BDB database; replaced by {`chainstate/`, `blocks/index/`, `blocks/revNNNNN.dat`<sup>[\[2\]](#note2)</sup>} in 0.8.0 | [PR #1677](https://github.com/bitcoin/bitcoin/pull/1677)
 `blk000?.dat`  | Block data (custom format, 2 GiB per file); replaced by `blocks/blkNNNNN.dat`<sup>[\[2\]](#note2)</sup> in 0.8.0 | [PR #1677](https://github.com/bitcoin/bitcoin/pull/1677)
 `addr.dat`     | Peer IP address BDB database; replaced by `peers.dat` in [0.7.0](https://github.com/bitcoin/bitcoin/blob/master/doc/release-notes/release-notes-0.7.0.md) | [PR #1198](https://github.com/bitcoin/bitcoin/pull/1198), [`928d3a01`](https://github.com/bitcoin/bitcoin/commit/928d3a011cc66c7f907c4d053f674ea77dc611cc)
+`onion_private_key` | Cached Tor onion service private key for `-listenonion` option. Was used for Tor v2 services; replaced by `onion_private_key_torv3` in [0.21.0](https://github.com/bitcoin/bitcoin/blob/master/doc/release-notes/release-notes-0.21.0.md) [PR #19954](https://github.com/bitcoin/bitcoin/pull/19954)
```
Missing column separator `|`?
```suggestion
`onion_private_key` | Cached Tor onion service private key for `-listenonion` option. Was used for Tor v2 services; replaced by `onion_private_key_torv3` in [0.21.0](https://github.com/bitcoin/bitcoin/blob/master/doc/release-notes/release-notes-0.21.0.md) | [PR #19954](https://github.com/bitcoin/bitcoin/pull/19954)
```

vasild replied:
Fixed

jonatack commented on doc/release-notes.md:
```diff
@@ -74,6 +74,15 @@ P2P and network changes
+  The private key for the Tor v3 service will be saved in `onion_private_key_torv3`.
+  To use the deprecated Tor v2 service (not recommended), then `onion_private_key`
+  can be copied over `onion_private_key_torv3`
+  (`cp -f onion_private_key onion_private_key_torv3`).
```
(`code`) markup inside parentheses; maybe easier to read as: 
```suggestion
  , e.g. `cp -f onion_private_key onion_private_key_torv3`.
```

vasild replied:
Done

jonatack commented on test/functional/p2p_addr_relay.py:
```diff
@@ -10,7 +10,7 @@
     NODE_NETWORK,
     NODE_WITNESS,
-    msg_addr,
+    msg_addrv2,
```
Should we test for both `msg_addr` and `msg`addrv2`?

vasild replied:
Right, `p2p_addr_relay.py` is the only test that exercises `msg_addr` and if we flip it to test `msg_addrv2` then `msg_addr` will be untested.

I tried to combine both in one test, but it became too convoluted and so I copied the new test to `p2p_addrv2_relay.py` and restored the original `p2p_addr_relay.py`.

jonatack commented:
Tested re-ACK bb01c77865a565764e4e2f86265afef45506d7d7

Reviewed the diff since my last review per `git range-diff 301959f 1550aa8 bb01c77`, re-reviewed the changeset, rebuilt, deleted the `onion_private_key_torv3` file from previous testing and relaunched; node seems to be running nominally. The additional tests are great. A few comments above to take or leave; none are blockers.

hebasto reviewed (commented):
Tested bb01c77865a565764e4e2f86265afef45506d7d7 on Linux Mint 20 (x86_64). I've verified P2P messages in the testnet with Wireshark 3.2.3.

- "master-..." means `bitcoind` compiled against master
- "pr-..." means `bitcoind` compiled against this PR
- a peer before connecting has no `peers.dat`

### Setup A -- master-peer connects to pr-node

The pr-node sends the "sendaddrv2" message to the master-peer just after sending the "verack" message.
In response to master-peer's "getaddr" message the pr-node sends the "addr" message with 1000 addresses and payload size 30003 bytes (1000*30+3).

### Setup B -- pr-peer connects to pr-node

Both the pr-node and the pr-peer send each other the "sendaddrv2" message just after sending the "verack" message.
In response to pr-peer's "getaddr" message the pr-node sends the "addrv2" message with 1000 addresses and payload size 16463 bytes. Some addresses I observed are 13 bytes long.

hebasto commented:
Does `addrv2` code follows https://github.com/bitcoin/bips/pull/967 ?

hebasto commented on src/net_processing.cpp:
```diff
@@ -2414,6 +2414,9 @@ void PeerManager::ProcessMessage(CNode& pfrom, const std::string& msg_type, CDat
         m_connman.PushMessage(&pfrom, CNetMsgMaker(greatest_common_version).Make(NetMsgType::VERACK));
 
+        // Signal ADDRv2 support (BIP155).
+        m_connman.PushMessage(&pfrom, CNetMsgMaker(INIT_PROTO_VERSION).Make(NetMsgType::SENDADDRv2));
```
76afce803bb5309a69517df5eac68529b7af2d29
```suggestion
        m_connman.PushMessage(&pfrom, CNetMsgMaker(greatest_common_version).Make(NetMsgType::SENDADDRv2));
```

vasild replied:
Done

hebasto commented on src/torcontrol.cpp:
```diff
@@ -532,8 +532,9 @@ void TorController::auth_cb(TorControlConnection& _conn, const TorControlReply&
         // Finally - now create the service
-        if (private_key.empty()) // No private key, generate one
-            private_key = "NEW:RSA1024"; // Explicitly request RSA1024 - see issue #9214
+        if (private_key.empty()) { // No private key, generate one
+            private_key = "NEW:ED25519-V3"; // Explicitly request key type - see issue #9214
+        }
```
bb01c77865a565764e4e2f86265afef45506d7d7

The comment about #9214 seems outdated now:
```suggestion
        if (private_key.empty()) {
            // No private key, generate one.
            private_key = "NEW:ED25519-V3";
        }
```

vasild replied:
The comment is still relevant - we want to explicitly request the key type in order to avoid problems like #9214 in the future.

In other words - we don't want somebody to change this to `NEW:BEST` because if/when Tor upgrades to v4 then we may have #9214 again.

hebasto replied:
Correct.

hebasto commented on doc/release-notes.md:
```diff
@@ -74,6 +74,15 @@ P2P and network changes
   node using P2P relay. This version reduces the initial broadcast guarantees
   for wallet transactions submitted via P2P to a node running the wallet. (#18038)
 
+- The Tor hidden service that is automatically created (`-listenonion`) will now
```
bb01c77865a565764e4e2f86265afef45506d7d7

See #19638:
```suggestion
- The Tor onion service that is automatically created (`-listenonion`) will now
```

vasild replied:
Changed

hebasto reviewed (approved):
ACK bb01c77865a565764e4e2f86265afef45506d7d7

May I suggest to replace s/"onion_private_key_torv3"/"onion_private_key_v3"/ as "tor" substring seems redundant here?

laanwj commented:
> May I suggest to replace s/"onion_private_key_torv3"/"onion_private_key_v3"/ as "tor" substring seems redundant here?

Kind of agree, the same came to mind for me but didn't bother to comment it initially. But a shorter filename that still contains the same information is good I suppose.

vasild commented:
Addressed review suggestions. Notice - I [restored](https://github.com/bitcoin/bitcoin/pull/19954#discussion_r498801055) the original version of `p2p_addr_relay.py` and the test for addrv2 is now in a new test `p2p_addrv2_relay.py`.

jonatack commented:
> I [restored](https://github.com/bitcoin/bitcoin/pull/19954#discussion_r498801055) the original version of `p2p_addr_relay.py` and the test for addrv2 is now in a new test `p2p_addrv2_relay.py`.

Nice!

hebasto reviewed (approved):
re-ACK c212c7eea6c0f67d7c65467c6c93f188aaeb02f3, only suggested changes since my [previous](https://github.com/bitcoin/bitcoin/pull/19954#pullrequestreview-500460244) review.

jonatack commented:
@vasild in the latest push, the `p2p_addr_relay.py` file perms were changed from 755 to 644; can you revert that?

jonatack commented:
@vasild in the latest push, the `p2p_addr_relay.py` file perms were changed from 755 to 644; could you please revert that?

<details><summary>a few suggestions for <code>p2p_addrv2_relay.py</code></summary><p>

```diff
diff --git a/test/functional/p2p_addrv2_relay.py b/test/functional/p2p_addrv2_relay.py
index 19cfd0eadb..ca593148cc 100755
--- a/test/functional/p2p_addrv2_relay.py
+++ b/test/functional/p2p_addrv2_relay.py
@@ -6,18 +6,17 @@
@@ -6,18 +6,17 @@
 Test addrv2 relay
 """
 
+import time
+
 from test_framework.messages import (
     CAddress,
+    msg_addrv2,
     NODE_NETWORK,
     NODE_WITNESS,
-    msg_addrv2,
 )
 from test_framework.p2p import P2PInterface
 from test_framework.test_framework import BitcoinTestFramework
-from test_framework.util import (
-    assert_equal,
-)
-import time
+from test_framework.util import assert_equal
 
 ADDRS = []
 for i in range(10):
@@ -48,11 +47,11 @@ class AddrReceiver(P2PInterface):
 
 class AddrTest(BitcoinTestFramework):
     def set_test_params(self):
-        self.setup_clean_chain = False
+        self.setup_clean_chain = True
         self.num_nodes = 1
 
     def run_test(self):
-        self.log.info('Create connection that sends addr messages')
+        self.log.info('Create connection that sends addrv2 messages')
         addr_source = self.nodes[0].add_p2p_connection(P2PInterface())
         msg = msg_addrv2()
 
@@ -61,7 +60,7 @@ class AddrTest(BitcoinTestFramework):
         with self.nodes[0].assert_debug_log(['addr message size = 1010']):
             addr_source.send_and_ping(msg)
 
-        self.log.info('Check that addr message content is relayed and added to addrman')
+        self.log.info('Check that addrv2 message content is relayed and added to addrman')
         addr_receiver = self.nodes[0].add_p2p_connection(AddrReceiver())
         msg.addrs = ADDRS
         with self.nodes[0].assert_debug_log([
```
</p></details>

jonatack commented on src/torcontrol.cpp:
```diff
@@ -718,7 +719,7 @@ void TorController::Reconnect()
 fs::path TorController::GetPrivateKeyFile()
 {
-    return GetDataDir() / "onion_private_key";
+    return GetDataDir() / "onion_private_key_v3";
```
```suggestion
    return GetDataDir() / "onion_v3_private_key";
```

vasild replied:
Done

jonatack commented on test/functional/p2p_invalid_messages.py:
```diff
@@ -22,7 +22,12 @@
+    hex_str_to_bytes,
+)
+import struct
+import time
```
nit: IIRC, per PEP8, OS-level and stdlib imports go at the top.

vasild replied:
Done

jonatack reviewed (commented):
Tested that:

- restarting a node after running `cp -f onion_private_key onion_private_key_v3` does indeed launch a Tor *v2* service

- following which, restarting again after `rm onion_private_key_v3` creates a new Tor *v3* service

I've also been testing this branch rebased over #20002, which is based on #19998, and is a clean rebase. 

A few more comments below. This PR seems close.

jonatack commented on src/net_processing.cpp:
```diff
@@ -2610,6 +2635,11 @@ void PeerManager::ProcessMessage(CNode& pfrom, const std::string& msg_type, CDat
 
+            // Skip invalid addresses, maybe a new BIP155 network id from the future.
+            if (!addr.IsValid()) {
+                continue;
```
Should skipping invalid BIP155 addresses be applied only for `addrv2` messages?
```cpp
if (!addr.IsValid() && msg_type == NetMsgType::ADDRv2) {
    continue;
}
```

vasild replied:
I realized that this code is not necessary so I dropped it altogether - an invalid address is not routable and is not reachable, so we are not going to do anything with it.

sipa commented on src/addrman.h:
```diff
@@ -264,6 +264,17 @@ friend class CAddrManTest;
 public:
+    //! Serialization versions.
+    enum class Format : uint8_t {
+        V0 = 0, //!< historic format, before commit e6b343d88
```
How about giving these a bit more descriptive names, like `V0_HISTORICAL`, `V1_DETERMINISTIC`, `V2_ASMAP`, `V3_BIP155`?

vasild replied:
Renamed

sipa commented on src/addrman.h:
```diff
@@ -318,8 +329,15 @@ friend class CAddrManTest;
+        if (m_format >= Format::V3) {
+            // Add ADDRV2_FORMAT to the version so that the CNetAddr and CAddress
+            // serialize methods produce an address in v2 format.
+            s.SetVersion(stream_version_orig | ADDRV2_FORMAT);
```
A cleaner approach is using `OverrideStream` here, which removes the need to undo the version change, or having a mutable version in CHashWriter.

vasild replied:
Right, changed to `OverrideStream`.

https://github.com/bitcoin/bitcoin/pull/19503 is also an alternative if/when it gets merged.

vasild commented:
Addressed review suggestions.

hebasto commented on src/addrman.h:
```diff
@@ -371,13 +390,25 @@ friend class CAddrManTest;
     }
 
     template<typename Stream>
```
pico-nit:
```suggestion
    template <typename Stream>
```

vasild replied:
Done

hebasto commented on src/addrman.h:
```diff
@@ -264,6 +264,17 @@ friend class CAddrManTest;
 public:
+    //! Serialization versions.
+    enum class Format : uint8_t {
+        V0_HISTORICAL = 0, //!< historic format, before commit e6b343d88
+        V1_DETERMINISTIC = 1, //!< for pre-asmap files
+        V2_ASMAP = 2, //!< for files including asmap version
+        V3_BIP155 = 3, //!< same as V2 plus addresses are in BIP155 format
```
pico-nit, as `clang-format` suggests, and s/V2/V2_ASMAP/ in comment:
```suggestion
        V0_HISTORICAL = 0,    //!< historic format, before commit e6b343d88
        V1_DETERMINISTIC = 1, //!< for pre-asmap files
        V2_ASMAP = 2,         //!< for files including asmap version
        V3_BIP155 = 3,        //!< same as V2_ASMAP plus addresses are in BIP155 format
```

vasild replied:
Done

hebasto commented on src/hash.h:
```diff
@@ -102,13 +102,14 @@ class CHashWriter
 
     const int nType;
-    const int nVersion;
+    int nVersion;
 public:
```
pico-nit:
```suggestion
    int nVersion;

public:
```

vasild replied:
Done

hebasto reviewed (approved):
re-ACK bbbc8e4c657e8ea81edb6aca2b1870cd3b7459ad.

hebasto commented:
I'd like that `s` and `s_` objects of the `Stream` class were named more descriptive, but I have no particular suggestion.

vasild commented:
Addressed review suggestions.

hebasto reviewed (approved):
re-ACK 703db15ae197c6b2e9489c14af441c5d796ff082, tested on Linux Mint 20 (x86_64).

Besides checking [network traffic](https://github.com/bitcoin/bitcoin/pull/19954#pullrequestreview-500239229), I've verified that my node is able to connect to peers with v2 and v3 onion addresses, and is reachable via v3 onion service. Created file with key in my datadir is `onion_v3_private_key`.

Thanks @vasild !

jonatack commented on src/hash.h:
```diff
@@ -102,13 +102,15 @@ class CHashWriter
 
     int GetType() const { return nType; }
     int GetVersion() const { return nVersion; }
+    void SetVersion(int n) { nVersion = n; }
```
03572e9 After `git diff c212c7e 703db15` using `OverrideStream`, it looks like we can now revert the changes in `hash.h` and `streams.h`.
```diff
+++ b/src/hash.h
@@ -102,7 +102,7 @@ private:
     const int nType;
-    int nVersion;
+    const int nVersion;
@@ -110,7 +110,6 @@ public:
     int GetVersion() const { return nVersion; }
-    void SetVersion(int n) { nVersion = n; }

+++ b/src/streams.h
@@ -603,7 +603,7 @@ class CAutoFile
     const int nType;
-    int nVersion;
+    const int nVersion; 
@@ -651,7 +651,6 @@ public:
     int GetVersion() const       { return nVersion; }
-    void SetVersion(int n)       { nVersion = n; }
```

vasild replied:
Good catch! Removed.

sipa commented on src/net_processing.cpp:
```diff
@@ -4104,14 +4137,14 @@ bool PeerManager::SendMessages(CNode* pto)
                     if (vAddr.size() >= MAX_ADDR_TO_SEND)
                     {
-                        m_connman.PushMessage(pto, msgMaker.Make(NetMsgType::ADDR, vAddr));
+                        m_connman.PushMessage(pto, msgMaker.Make(msg_type, vAddr));
```
If you use the `CNetMsgMaker:: Make(int nFlags, std::string msg_type, Args&&... args)` variant here, you can control the serialization flags, without needing the special casing for ADDRV2 inside `Make`.

vasild replied:
A good interface is [hard or impossible to misuse](https://testing.googleblog.com/2018/07/code-health-make-interfaces-hard-to.html). If we add `ADDRV2_FORMAT` to `nFlags` inside `Make()` that makes it impossible for the caller to misuse it like:

```cpp
msgMaker.Make(0, NetMsgType::ADDRv2, vAddr); // huh!?
```

sipa replied:
@vasild I agree with the notion, but this is just the wrong layer to deal with that. The serialization framework (which I'd consider netmessagemaker to be a thin wrapper around) shouldn't have knowledge about the data being serialized, or even about which message types exist.

> I think it is not good to impose on the caller: "whenever you pass `NetMsgType::ADDRv2` for `msg_type` you must also pass `ADDRV2_FORMAT` in `nFlags`".

I think that's the wrong way to think about it. `CNetMsgMaker` should do as it's told: serializing the object passed with the specified flags, and nothing else. Knowledge about what message types exist, and how data is serialized in them belongs in net_processing.

vasild replied:
That is also a valid point. Changed.

vasild commented on src/addrman.h:
```diff
@@ -370,14 +389,26 @@ friend class CAddrManTest;
-        unsigned char nVersion;
-        s >> nVersion;
+
+        uint8_t format_tmp;
```
Used `CustomUintFormatter` here, but in `Serialize` it did not compile:
```
src/serialize.h:504:15: error: invalid operands to binary expression ('CAddrMan::Format' and 'int')
        if (v < 0 || v > MAX) throw std::ios_base::failure("CustomUintFormatter value out of range");
            ~ ^ ~
```
so I left the `static_cast` there.

vasild commented on src/streams.h:
```diff
@@ -650,6 +651,7 @@ class CAutoFile
     //
     int GetType() const          { return nType; }
     int GetVersion() const       { return nVersion; }
+    void SetVersion(int n)       { nVersion = n; }
```
Removed

vasild commented on src/addrman.h:
```diff
@@ -264,6 +264,17 @@ friend class CAddrManTest;
+    };
+
+    //! Serialization / unserialization format.
+    Format m_format = Format::V3_BIP155;
```
I agree "it's fine to just always serialize in the latest format", changed the code accordingly - now the format variable is a local one in `Unserialize()` while `Serialize()` just serializes the constant `Format::V3_BIP155`.

This is also closer to the behavior before this PR (it always serialized in version 2, the latest at the time).

hebasto replied:
Does this mean that `peers.dat` becomes unreadable if a user choose to downgrade his node at some point in the future?

jonatack replied:
Testing this branch, restarting after `cp onion_private_key onion_v3_private_key` to run a v2 service, and then a second time when restarting after `rm onion_v3_private_key` to return to v3, I'm seeing `2020-10-06T16:41:46Z Invalid or missing peers.dat; recreating` messages.

jonatack replied:
Testing this branch:

A) One or two restarts after `cp onion_private_key onion_v3_private_key` to run a v2 service

```
ERROR: DeserializeDB: Deserialize or I/O error - ReadCompactSize(): size too large: iostream error 
Invalid or missing peers.dat; recreating
```

B) Eventually on restart

`Loaded 305 addresses from peers.dat`

C) One or two restarts after `rm onion_v3_private_key` to return to v3

```
ERROR: DeserializeDB: Deserialize or I/O error - ReadCompactSize(): size too large: iostream error 
Invalid or missing peers.dat; recreating
```

D) Eventually on restart:

`Loaded 297 addresses from peers.dat`

I repeated this a few times.

vasild replied:
> Does this mean that peers.dat becomes unreadable if a user choose to downgrade his node at some point in the future?

Yes. We can't have bitcoind-before-this-PR to read e.g. torv3 addresses from `peers.dat`.

hebasto replied:
Does this require mention in release notes?

vasild commented:
Addressed review suggestions and extended some tests.

vasild commented:
[Filtered code coverage report](https://people.freebsd.org/~vd/pr19954_946b79a9d_coverage/) (files not modified by this PR are omitted and not modified lines in files that are otherwise modified are dimmed).

[List of modified and not covered lines](https://people.freebsd.org/~vd/pr19954_946b79a9d_coverage/modified_and_not_covered.html).

sipa commented on src/net_processing.cpp:
```diff
@@ -2582,16 +2588,37 @@ void PeerManager::ProcessMessage(CNode& pfrom, const std::string& msg_type, CDat
-            Misbehaving(pfrom.GetId(), 20, strprintf("addr message size = %u", vAddr.size()));
+            Misbehaving(pfrom.GetId(), 20,
+                        strprintf("%s message size = %u",
+                                  msg_type == NetMsgType::ADDR ? "addr" : "addrv2",
```
This can instead be:

```c++
Misbehaving(pfrom.GetId(), 20, strprintf("%s message size = %u", msg_type, vAddr.size()));
```

As msg_type is just `const char*`.

vasild replied:
Changed. Btw `msg_type` is `std::string` (not `const char*`). `strprintf()` takes it either way.

sipa commented on src/protocol.h:
```diff
@@ -76,6 +76,18 @@ extern const char* VERACK;
+ * like the addr message, but is extended to allow gossiping of longer node
+ * addresses (see BIP155).
+ */
+extern const char *ADDRv2;
```
Style nit: constants are all caps, so `ADDRV2`

vasild replied:
Renamed

sipa commented on src/protocol.h:
```diff
@@ -76,6 +76,18 @@ extern const char* VERACK;
+ * It also implies that its sender can encode as ADDRv2 and would send ADDRv2
+ * instead of ADDR to a peer that has signaled ADDRv2 support by sending SENDADDRv2.
+ */
+extern const char *SENDADDRv2;
```
Style nit: constants are all caps, so `SENDADDRV2`.

vasild replied:
Renamed

sipa commented:
Code review ACK modulo nits.

I tested:
* `externalip=` with torv3 address, and saw it reported in logs
* upgrading a automatically generated torv2 address to a v3 one
* automatically creating a torv3 address
* connecting two nodes with this patch to each other, and seeing `addrv2` reported in `getpeerinfo` output

vasild commented on src/net_processing.cpp:
```diff
@@ -2582,16 +2588,37 @@ void PeerManager::ProcessMessage(CNode& pfrom, const std::string& msg_type, CDat
+
+        try {
+            s >> vAddr;
+        } catch (const std::exception& e) {
```
I agree, removed the `try/catch`.

vasild commented:
Addressed review suggestions.

laanwj commented:
re-ACK 1f06a2901d1ee924b95f482e8693842ee07863de
Updated my nodes with the old version of this patch to the current one (and renamed the key file, the new name is good), it still works as expected.

vasild commented:
Just to clarify (assuming this PR goes in 0.21): the format of `peers.dat` is changed in a backwards-incompatible way. Meaning that 0.20 and older will be unable to read `peers.dat` written by 0.21 or newer. This boils down to the fact that 0.20 and older cannot read e.g. torv3 addresses encoded in BIP155-addrv2 format.

This is not related in any way to the `onion_private_key` and `onion_v3_private_key` files and whether the node creates an onion service or not.

jonatack commented on doc/release-notes.md:
```diff
@@ -74,6 +74,15 @@ P2P and network changes
+  The private key for the Tor v3 service will be saved in `onion_v3_private_key`.
+  To use the deprecated Tor v2 service (not recommended), then `onion_private_key`
+  can be copied over `onion_v3_private_key`, e.g.
+  `cp -f onion_private_key onion_v3_private_key`.
```
1f06a2901d1ee924b95f482e8693842ee07863de Perhaps inform users of the debug log messages, e.g. `ERROR: DeserializeDB: Deserialize or I/O error` and `Invalid or missing peers.dat`, that they may see if they use Tor v2 henceforth.

vasild replied:
Added a release notes mention.

dongcarl reviewed (commented):
Code review ACK 1f06a2901d1ee924b95f482e8693842ee07863de

jonatack reviewed (commented):
Tested re-ACK 1f06a2901d1ee924b95f482e8693842ee07863de modulo comments below. Started new v3 service by deleting `onion_v3_private_key` and after running previously on master, no errors, local addr new tor v3 address. Restarted after `cp onion_private_key onion_v3_private_key`, saw `ERROR: DeserializeDB: Deserialize or I/O error - ReadCompactSize(): size too large: iostream error` and `Invalid or missing peers.dat; recreating`, local addr usual tor v2 address. Restarted after deleting `onion_v3_private_key`, no errors, local addr new tor v3 address. I see "addrv2" and "sendaddrv2" messages in getpeerinfo in both the "bytesrecv_per_msg" and "bytessent_per_msg" objects. Would be good to nail down those errors.

jonatack commented:
> Just to clarify (assuming this PR goes in 0.21): the format of `peers.dat` is changed in a backwards-incompatible way. Meaning that 0.20 and older will be unable to read `peers.dat` written by 0.21 or newer. This boils down to the fact that 0.20 and older cannot read e.g. torv3 addresses encoded in BIP155-addrv2 format.

Indeed, it seems we need to be clear about informing people of this backward incompatibility with regards to peers.dat. Empirically, it looks like it's related to using the v2 private key but it may not be the actual case.

sipa committed:
Support bypassing range check in ReadCompactSize

This is needed when we want to encode an arbitrary number as CompactSize
like node service flags, which is a bitmask and could be bigger than the
usual size of an object.

laanwj commented on doc/release-notes.md:
```diff
@@ -74,6 +74,15 @@ P2P and network changes
+  for Tor v2 (if any) will be left untouched in `onion_private_key` in `-datadir`
+  and can be removed if not needed. Bitcoin Core will no longer attempt to read it.
+  The private key for the Tor v3 service will be saved in `onion_v3_private_key`.
+  To use the deprecated Tor v2 service (not recommended), then `onion_private_key`
+  can be copied over `onion_v3_private_key`, e.g.
+  `cp -f onion_private_key onion_v3_private_key`.
```
Personally, I think it would be ok to not support v2 addresses at all anymore. Given the short time they have left. But that there's a workaround (even if clunky) might help for some people.

vasild replied:
@laanwj, I agree, given torv2 addresses will cease in less than a year - no need to add new config options (or extend existent ones) only to deprecate them immediately.

laanwj commented:
> Just to clarify (assuming this PR goes in 0.21): the format of peers.dat is changed in a backwards-incompatible way. Meaning that 0.20 and older will be unable to read peers.dat written by 0.21 or newer. This boils down to the fact that 0.20 and older cannot read e.g. torv3 addresses encoded in BIP155-addrv2 format.

This only means starting with (effectively) an empty peers.dat if you downgrade, right? That's okay. Loss of peers.dat is not a big deal to most users. As long as it doesn't result in the program *not starting at all*. That would be kind of unfriendly.

sipa commented:
Earlier today on IRC:

```
13:39:37 < vasild> CAddress::unser: time=1601829922, services=134218813, addr=62.107.200.30:8333
13:39:56 < vasild> CAddress::unser: time=1600339981, services=134217741, addr=akinbo7tlegsnsxn.onion:8333
13:40:31 < vasild> jonatack: wumpus: so...
13:41:05 < vasild> in the network there are some peers with services=unusually high value
13:42:07 < sipa> that's 0x0800000d
13:42:14 < vasild> they get in via p2p gossip into addrman without causing problems
13:42:16 < vasild> sipa: right
13:43:18 < vasild> using addrv1
13:44:03 < vasild> then addrman saves it on disk just fine in V3_BIP155 format using READWRITE(COMPACTSIZE(services_tmp))
13:44:44 < vasild> however, on read back from disk it bricks because ReadCompactSize() contains this at the end:
13:44:53 < vasild>     if (nSizeRet > (uint64_t)MAX_SIZE)
13:44:54 < vasild>         throw std::ios_base::failure("ReadCompactSize(): size too large");
13:45:18 < vasild> and MAX_SIZE is quite low value 0x02000000
13:45:48 < vasild> I wondered before wtf is that limit but decided it is harmless
13:46:14 < sipa> vasild: ehhhh
13:46:27 < sipa> this means we need to bypass that limit when deserializing addrv2
13:46:46 < vasild> so WriteCompactSize() can write >0x02000000 but ReadCompactSize() cannot read it back :-X
```

I wrote a small patch that allows bypassing the range check in ReadCompactSize: https://github.com/sipa/bitcoin/commits/202010_compactsize_bound. Feel free to cherry-pick or squash. It allows using without range check using `Using<CompactSizeFormatter<false>>(obj)` instead of `COMPACTSIZE(obj)`.

laanwj commented:
> 13:46:46 < vasild> so WriteCompactSize() can write >0x02000000 but ReadCompactSize() cannot read it back :-X

Ouch, good catch! This makes me doubt the choice to use a `CompactSize` for anything else than a size a little. It seems like misuse of a typed value. @sipa's patch to allow to bypass the check would solve the immediate issue, I just hope it won't cause problems for maintainers in the future again if other sanity checks are added, with people thinking they'll be size specific.

sipa commented:
@laanwj Well, they're already used as non-size values in BIP152, though luckily they're restricted to 16-bit values there, so the range limit was not an issue.

I think it's fine to use it in BIP155 as well. The code is a bit confusing but perhaps we should just rename CompactSize to something else at some point. It's already known under other names outside Bitcoin Core anyway.

jonatack commented:
> This only means starting with (effectively) an empty peers.dat if you downgrade to a pre-0.21 version, right?

Yes, that's what I'm seeing.

> (Nevertheless, it does make sense to mention the `peers.dat` format change and consequences in the release notes)

:+1:

vasild commented on src/streams.h:
```diff
@@ -60,6 +60,7 @@ class OverrideStream
     int GetVersion() const { return nVersion; }
     int GetType() const { return nType; }
     size_t size() const { return stream->size(); }
+    void ignore(size_t size) { return stream->ignore(size); }
```
The `ignore()` modifies the state of the stream. It is not `const` in `CDataStream` and `CAutoFile` so this would not compile.

jonatack replied:
It does compile, though (I test if my suggestions compile before making them.)

jonatack replied:
It does compile (I test if my suggestions compile before making them), though arguably const on a void function may not make sense. Consider this suggestion a question.

jonatack replied:
"void f() const makes the function itself const. This only really has meaning for member functions. Making a member function const means that it cannot call any non-const member functions, nor can it change any member variables. It also means that the function can be called via a const object of the class." http://www.cplusplus.com/forum/general/12087/

vasild replied:
I think it should not be `const` because it modifies the state of the stream. I guess it is not `const` for the same reason in the other stream classes.

vasild commented on src/netaddress.cpp:
```diff
@@ -474,6 +474,26 @@ bool CNetAddr::IsInternal() const
+    }
+
+    return false;
+}
```
Done

vasild commented:
Addressed review suggestions plus

* Added release notes on the `peers.dat` format change
* Made it possible to read big compact-size numbers (thanks to @sipa for the patch) this will enable addrman to read arbitrary big service flags from its database and also for the p2p layer to receive them when in `addrv2` encoding (the old `addr` encoding uses fixed 8-byte representation)
* Added a human-readable error message if addrman detects a future format of `peers.dat`

jonatack commented:
Initial updated suggestion for the release notes, though I am sure it can be improved further (pinging @harding). It may need to be edited by hand in the wiki if the notes have already been merged and branched off.

```
Migration from deprecated Tor v2 to Tor v3
-----------------------

- The Tor onion service that is automatically created (`-listenonion`) will now
  be created as a Tor v3 service instead of Tor v2. The private key that was used
  for Tor v2 (if any) will be left untouched in `onion_private_key` in `-datadir`
  and can be removed if not needed. Bitcoin Core will no longer attempt to read it.
  The private key for the Tor v3 service will be saved in `onion_v3_private_key`.
  To use the deprecated Tor v2 service (not recommended), then `onion_private_key`
  can be copied over `onion_v3_private_key`, e.g.
  `cp -f onion_private_key onion_v3_private_key`. (#19954)

Peers.dat file format change
-----------------------

The node's known peers are persisted to disk in a file called
`peers.dat`. Starting with this v0.21.0 release, the `peers.dat` file format is
changed in a backwards-incompatible way to accommodate the storage of larger
BIP155 addresses, such as Tor v3 addresses. Pre-0.21 Bitcoin Core software will
not be able to read this new `peers.dat` format. In the event of downgrading to
pre-0.21 after running 0.21, the debug log may print an error that
deserialization has failed and a message that peers.dat is invalid or missing
and is being recreated, and bitcoind will continue normal operation with a new
`peers.dat` file. (#19954)
```

jonatack commented on src/addrman.h:
```diff
@@ -370,14 +382,34 @@ friend class CAddrManTest;
+        if (format > maximum_supported_format) {
+            throw std::ios_base::failure(strprintf(
+                "Unsupported format of addrman database: %u. Maximum supported is %u. "
+                "Continuing operation without using the saved list of peers.",
```
6fce3ce maybe
```suggestion
                "Continuing operation with a new peers.dat in the maximum supported format",
```
(and without the full stop), as currently it would print

`ERROR: DeserializeDB: Deserialize or I/O error - Unsupported format of addrman database: 4. Maximum supported is 3. Continuing operation without using the saved list of peers.: iostream error`

vasild replied:
"Continuing operation by creating a new peers.dat" is, strictly speaking, not true because the new `peers.dat` will not be created at the time this message is printed. It will be created later during shutdown or during one of the periodic dumps to disk.

I left it as is.

jonatack reviewed (commented):
ACK 9b56a68a85ce32b13d3db605ac56bb9743437d2f per `git diff 1f06a29 9b56a68`, tested with v3/v2/v3 onion private keys, tested on v0.18.1, and tested the new error message.

In my testing, the new patch has fixed the DB deserialize errors and peers.dat messages on this branch after `cp onion_private_key onion_v3_private_key`.

Restarting on v0.18.1, as expected I see `ERROR: DeserializeDB: Deserialize or I/O error - CAutoFile::read: end of file: iostream error` and `Invalid or missing peers.dat; recreating`.

vasild commented on src/serialize.h:
```diff
@@ -24,7 +24,11 @@
-static const unsigned int MAX_SIZE = 0x02000000;
+/**
+ * The maximum size of a serialized object in bytes or number of elements
+ * (for eg vectors) when the size is encoded as CompactSize.
```
The dot terminates the doxygen title.
![dox1](https://user-images.githubusercontent.com/266751/95575574-3abae680-0a2f-11eb-91e7-8b39ef12e317.jpg)
![dox2](https://user-images.githubusercontent.com/266751/95575577-3b537d00-0a2f-11eb-98df-659dd55565dc.jpg)

jonatack replied:
TIL, thanks. Apparently, per https://www.doxygen.nl/manual/docblocks.html#docexamples: "To enable this behavior you should set JAVADOC_AUTOBRIEF to YES in the configuration file. If you enable this option and want to put a dot in the middle of a sentence without ending it, you should put a backslash and a space after it. Here is an example: `/** Brief description (e.g.\ using only a few words). Details follow. */`"

Indeed, doesn't seem worth it.

harding commented:
@jonatack The peers.dat text looks fine to me.  Here's a few suggestions for the v3 onion service text, but I also think that was fine.

```diff
-The Tor onion service that is automatically created (`-listenonion`) will now be
+The Tor onion service that is automatically created by setting the `-listenonion` configuration parameter will now be
 created as a Tor v3 service instead of Tor v2. The private key that was used for
-Tor v2 (if any) will be left untouched in `onion_private_key` in `-datadir` and
+Tor v2 (if any) will be left untouched in the `onion_private_key` file in the data directory (see `-datadir`)  and
 can be removed if not needed. Bitcoin Core will no longer attempt to read it.
-The private key for the Tor v3 service will be saved in `onion_v3_private_key`.
+The private key for the Tor v3 service will be saved in a file named `onion_v3_private_key`.
 To use the deprecated Tor v2 service (not recommended), then `onion_private_key`
 can be copied over `onion_v3_private_key`, e.g. `cp -f onion_private_key
 onion_v3_private_key`. (#19954)
```

jonatack commented:
Thanks @harding!

vasild commented:
Did some rewording of the release notes, as suggested.

jonatack commented on doc/release-notes.md:
```diff
@@ -60,6 +60,14 @@ From Bitcoin Core 0.20.0 onwards, macOS versions earlier than 10.12 are no
+the file is modified by 0.21.0 or newer then older versions will not be able to
+read it. Those old versions, in the event of a downgrade, will log an error
+message that deserialization has failed and will continue normal operation
+as if the file was missing, creating a new empty one. (#19954)
```
- Missing header
- can omit "in order"
- missing comma s/newer then older/newer, then older/ ...though this formulation of newer/older/old seems awkward and is also less clear

Sjors commented:
Tested that spinning up a v3 onion and connecting to it still works as of dcf0cb477699d11afd0ff37c8bfb2b1b4f7f1ee5. The PR title should probably mention that we start using `sendaddrv2` here, with TorV3 just being the cherry on top.

ariard reviewed (commented):
Ongoing review, minor points so far.

I do have a concern with a BIP about future extensions of the network IDs table. 

How a BIP-155-with-extended-table node will learn the set of supported network IDs by a BIP-155-with-original-table peer and thus avoid to waste bandwidth on non-understood by receiver addresses ? 

I think this is independent of the question if a connection is established or not to the known network. Maybe `sendaddrv2` should have a version bit, this would avoid introducing `sendaddrvX` at each table update, as far as yet-to-added network address are under maximum v2 length.

vasild commented on src/addrman.h:
```diff
@@ -370,14 +382,34 @@ friend class CAddrManTest;
+        }
+
+        int stream_version = s_.GetVersion();
+        if (format >= Format::V3_BIP155) {
```
Yes, this code will now only be reached if `format == V3_BIP155`. In the future `format` may be higher. By using `>=` we assume that a future format `V4_FOO` will also use `addrv2` encoding in which case this code will not have to be changed. If a future `V4_FOO` does not use `addrv2`, then this line will have to be changed to `==`.

Notice however that if we use `==` now and a future `V4_FOO` still uses `addrv2`, then this line will have to be changed from `==` to `>=`.

vasild commented on src/net.h:
```diff
@@ -868,6 +868,11 @@ class CNode
     bool m_limited_node{false}; //after BIP159, set by version message
+    /**
+     * Whether the peer has signaled support for receiving ADDRv2 (BIP155)
+     * messages, implying a preference to receive ADDRv2 instead of ADDR ones.
```
Do you mean this?

> Once `B` receives `sendaddrv2` from `A`:
>
> * `B` MUST use `addrv2` messages to gossip addresses to `A`
> * `B` MUST accept `addrv2` messages from `A`
> * `B` MAY accept `addr` messages from `A`

The current PR accepts both `addr` and `addrv2` no matter whether `sendaddrv2` has been sent or received.

ariard replied:
> Do you mean this?

Yes, I think the BIP should state this clearly to hint other implementations. The current PR code is correct.

vasild commented on src/net.h:
```diff
@@ -1115,11 +1120,16 @@ class CNode
+        // Whether the peer supports the address in `_addr`. For example,
+        // nodes that do not implement BIP155 cannot receive Tor v3 addresses
+        // because they require ADDRv2 (BIP155) encoding.
+        const bool addr_format_supported = m_wants_addrv2 || _addr.IsAddrV1Compatible();
```
`PushAddress()` is called from 5 places. We don't want to duplicate the same code in 5 places.

vasild replied:
`PushAddress()` is called from 5 places. We don't want to duplicate the same code in 5 places.

ariard replied:
Well another alternative is a new method `IsAddrUnderstood` on `CNode` making the call to `PushAddress()` conditional on it results. It can be done later, not a real concern.

jnewbery replied:
`PushAddress()` and all of the addr relay data will eventually move into net processing. See the commits in https://github.com/jnewbery/bitcoin/commits/2020-06-cs-main-split. That work obviously shouldn't hold up new features like BIP 155 and I'll continue to rebase those commits on master as things change. #19829 and #19910 are the next PRs to review if you'd like to help.

vasild commented on src/net_processing.cpp:
```diff
@@ -2582,16 +2588,25 @@ void PeerManager::ProcessMessage(CNode& pfrom, const std::string& msg_type, CDat
+        if (msg_type == NetMsgType::ADDRV2) {
+            // Add ADDRV2_FORMAT to the version so that the CNetAddr and CAddress
+            // unserialize methods know that an address in v2 format is coming.
+            stream_version |= ADDRV2_FORMAT;
```
`addr` and `addrv2` are just a transport encodings. Once received, the address is not in any particular encoding, it is a `CAddress` object in the memory of the node, not in `addr` nor in `addrv2` encoding.

So, when we further relay a `CAddress` object we don't care in what encoding we received it.

ariard replied:
> So, when we further relay a CAddress object we don't care in what encoding we received it.

Right, it's a just a matter of stating this in the BIP to avoid other implementations not propagating addresses due to receiving transport encoding A but forgetting to relay the decoded addr stream to peers speaking transport encoding B.

sipa replied:
I don't think we should aim to specify relay behavior in BIPs. There is no requirement that all software behaves the same, and best practices around it are constant changing in response to attacks and observed network conditions.

The BIP specifies what means of communicating addresses are available. What that gets used for is local node policy.

hebasto reviewed (approved):
re-ACK dcf0cb477699d11afd0ff37c8bfb2b1b4f7f1ee5, the node works flawlessly in all of the modes: Tor-only, clearnet-only, mixed.

sipa commented:
ACK dcf0cb477699d11afd0ff37c8bfb2b1b4f7f1ee5

jonatack commented:
re-ACK dcf0cb477699d11afd0ff37c8bfb2b1b4f7f1ee5 per `git diff 9b56a68 dcf0cb4` only change since last review is an update to the release notes which partially picked up the suggested text

jonatack commented:
re-ACK dcf0cb477699d11afd0ff37c8bfb2b1b4f7f1ee5 per `git diff 9b56a68 dcf0cb4` only change since last review is an update to the release notes which partially picked up the suggested text.

Note: since blowing out my peers.dat file while testing this branch, with a mixed clearnet/onion node I haven't seen any ipv6 connections at all, and no inbound onions yet, for two days so far.

ariard commented:
Code Review ACK dcf0cb4

MarcoFalke commented on src/test/netbase_tests.cpp:
```diff
@@ -443,4 +447,105 @@ BOOST_AUTO_TEST_CASE(netbase_dont_resolve_strings_with_embedded_nul_characters)
+        0x83766279U /* Tue Nov 22 11:22:33 UTC 2039 */
+    ),
+    CAddress(
+        CService(CNetAddr(in6addr_loopback), 0xf1f2 /* port */),
```
```
+ make -C src --jobs=1 check-security V=1
make: Entering directory '/bitcoin/distsrc-x86_64-linux-gnu/src'
Checking binary security...
READELF=/gnu/store/pzq9s5ldl90h72p26jkdakyccdg9ib3k-profile/bin/x86_64-linux-gnu-readelf OBJDUMP=x86_64-linux-gnu-objdump OTOOL= /gnu/store/pzq9s5ldl90h72p26jkdakyccdg9ib3k-profile/bin/python3.7 ../contrib/devtools/security-check.py bitcoind  bitcoin-cli bitcoin-tx bitcoin-wallet test/test_bitcoin bench/bench_bitcoin qt/bitcoin-qt  
make: Leaving directory '/bitcoin/distsrc-x86_64-linux-gnu/src'
+ case "$HOST" in
+ make -C src --jobs=1 check-symbols V=1
make: Entering directory '/bitcoin/distsrc-x86_64-linux-gnu/src'
Checking glibc back compat...
READELF=/gnu/store/pzq9s5ldl90h72p26jkdakyccdg9ib3k-profile/bin/x86_64-linux-gnu-readelf CPPFILT=/gnu/store/pzq9s5ldl90h72p26jkdakyccdg9ib3k-profile/bin/x86_64-linux-gnu-c++filt /gnu/store/pzq9s5ldl90h72p26jkdakyccdg9ib3k-profile/bin/python3.7 ../contrib/devtools/symbol-check.py bitcoind  bitcoin-cli bitcoin-tx bitcoin-wallet test/test_bitcoin bench/bench_bitcoin qt/bitcoin-qt  
test/test_bitcoin: export of symbol in6addr_loopback not allowed
test/test_bitcoin: failed EXPORTED_SYMBOLS
make: *** [Makefile:20952: check-symbols] Error 1
make: Leaving directory '/bitcoin/distsrc-x86_64-linux-gnu/src'

MarcoFalke replied:
This breaks the guix/gitian build. cc @fanquake

vasild replied:
The patch below may fix it, but I can't confirm because on my machine `contrib/devtools/symbol-check.py` is happy with `src/test/test_bitcoin` which contains:

```sh
nm src/test/test_bitcoin |grep in6addr_loopback
                 U in6addr_loopback
```

<details>
<summary>possible workaround symbol-check.py in6addr_loopback</summary>

```diff
diff --git i/src/test/netbase_tests.cpp w/src/test/netbase_tests.cpp
index eb0d95a37..f473c092a 100644
--- i/src/test/netbase_tests.cpp
+++ w/src/test/netbase_tests.cpp
@@ -447,25 +447,27 @@ BOOST_AUTO_TEST_CASE(netbase_dont_resolve_strings_with_embedded_nul_characters)
     BOOST_CHECK(!LookupSubNet(std::string("5wyqrzbvrdsumnok.onion\0example.com\0", 35), ret));
 }
 
 // Since CNetAddr (un)ser is tested separately in net_tests.cpp here we only
 // try a few edge cases for port, service flags and time.
 
+static const struct in6_addr in6addr_loopback_ = IN6ADDR_LOOPBACK_INIT;
+
 static const std::vector<CAddress> fixture_addresses({
     CAddress(
-        CService(CNetAddr(in6addr_loopback), 0 /* port */),
+        CService(CNetAddr(in6addr_loopback_), 0 /* port */),
         NODE_NONE,
         0x4966bc61U /* Fri Jan  9 02:54:25 UTC 2009 */
     ),
     CAddress(
-        CService(CNetAddr(in6addr_loopback), 0x00f1 /* port */),
+        CService(CNetAddr(in6addr_loopback_), 0x00f1 /* port */),
         NODE_NETWORK,
         0x83766279U /* Tue Nov 22 11:22:33 UTC 2039 */
     ),
     CAddress(
-        CService(CNetAddr(in6addr_loopback), 0xf1f2 /* port */),
+        CService(CNetAddr(in6addr_loopback_), 0xf1f2 /* port */),
         static_cast<ServiceFlags>(NODE_WITNESS | NODE_COMPACT_FILTERS | NODE_NETWORK_LIMITED),
         0xffffffffU /* Sun Feb  7 06:28:15 UTC 2106 */
     )
 });
 
 // fixture_addresses should equal to this when serialized in V1 format.
```

sipa replied:
See #19954.

vasild commented:
Congratulations to everybody involved :tada: 

Notice - I did not write the BIP, did not write the code initially, did not do the reviews nor did I test it much or nag people to prioritize it. Yet, this got a clean specification, solid code, thorough reviews and testing and was merged in time for 0.21. A true team effort!

Just Epic!

jnewbery commented on doc/release-notes.md:
```diff
@@ -60,6 +60,14 @@ From Bitcoin Core 0.20.0 onwards, macOS versions earlier than 10.12 are no
 
+The node's known peers are persisted to disk in a file called `peers.dat`. The
+format of this file has been changed in a backwards-incompatible way in order to
+accommodate the storage of Tor v3 and other BIP155 addresses. This means that if
+the file is modified by 0.21.0 or newer then older versions will not be able to
+read it. Those old versions, in the event of a downgrade, will log an error
+message that deserialization has failed and will continue normal operation
+as if the file was missing, creating a new empty one. (#19954)
```
"if the file is modified by 0.21.0 or newer then older versions will not be able to read it. Those old versions, in the event of a downgrade, will log an error message that deserialization has failed and will continue normal operation"

I don't see where this is enforced in 0.20 or earlier versions. If I'm understanding the v0.20 code correctly, then it'll try to unserialize the file, read garbage for the CAddress objects because the format has changed, but then not necessarily fail.

vasild replied:
You are right!

A dirty hack like the following would trick `<=0.20.1` to always fail deserialization of files in format `>=Format::V3_BIP155` with `Incorrect keysize in addrman deserialization`

```diff
diff --git i/src/addrman.h w/src/addrman.h
index b4089dc89..9f6614cd1 100644
--- i/src/addrman.h
+++ w/src/addrman.h
@@ -326,19 +326,19 @@ public:
     void Serialize(Stream& s_) const
     {
         LOCK(cs);
 
         // Always serialize in the latest version (currently Format::V3_BIP155).
 
         OverrideStream<Stream> s(&s_, s_.GetType(), s_.GetVersion() | ADDRV2_FORMAT);
 
         s << static_cast<uint8_t>(Format::V3_BIP155);
-        s << ((unsigned char)32);
+        s << ((unsigned char)33);
         s << nKey;
         s << nNew;
         s << nTried;
 
         int nUBuckets = ADDRMAN_NEW_BUCKET_COUNT ^ (1 << 30);
         s << nUBuckets;
         std::map<int, int> mapUnkIds;
         int nIds = 0;
         for (const auto& entry : mapInfo) {
@@ -406,18 +406,21 @@ public:
             // Add ADDRV2_FORMAT to the version so that the CNetAddr and CAddress
             // unserialize methods know that an address in addrv2 format is coming.
             stream_version |= ADDRV2_FORMAT;
         }
 
         OverrideStream<Stream> s(&s_, s_.GetType(), stream_version);
 
         unsigned char nKeySize;
         s >> nKeySize;
+        if (format >= Format::V3_BIP155) {
+            --nKeySize;
+        }
         if (nKeySize != 32) throw std::ios_base::failure("Incorrect keysize in addrman deserialization");
         s >> nKey;
         s >> nNew;
         s >> nTried;
         int nUBuckets = 0;
         s >> nUBuckets;
         if (format >= Format::V1_DETERMINISTIC) {
             nUBuckets ^= (1 << 30);
         }
```

jnewbery replied:
Yes, that would work. It's essentially repurposing the keysize field in the peers.dat file as a "non forward-compatible version number" (i.e. if a newer version of the software writes a higher number here, then an older version of the software wouldn't be able to read that file).

I believe that's what you wanted to achieve with the code here: https://github.com/bitcoin/bitcoin/pull/19954/files#diff-164bd9e2e30f54d0a79eb7cc372309e2f2155edc6c3f051290ab078f03f6a771R396, but of course that only applies for v0.21 onwards, for file versions higher than `Format::V3_BIP155`. That code means that we now have no way to make a downgrade-safe change to the peers.dat file version. Any bump to the format number means that older software versions can't read the file. Note that we have made downgrade-safe version changes before (I think both v0-v1 and v1-v2 were both downgrade-safe), and it would be nice to keep this as a possibility.

Repurposing the keysize field and reverting the change to the format field seems ok and would almost be turning them into major (non-forward-compatible) and minor (forward-compatible) version numbers. It should go without saying that we always want to have backward-compatibility, so peers.dat files written by old versions are readable by new versions.

Obviously we should get any changes made before the 0.21 release.

vasild replied:
See if this will make sense to you: https://github.com/bitcoin/bitcoin/pull/20284
